### ISSUE TITLE
To add capability use NSUserDefaults with appGroup

### ISSUE THREAD

**Issue by shenmueXVI** [mentrena/SyncKit#16]
Hi, your library is amazing, thank you. I use appgroup userDefaults in my app for widget, watch extension and main app. My CoreData is stored in appgroup container. And was bug, when I sync data for main app, widget, watch extension - as dublicate actions (add data or delete data). I found problem - deviceIdentifier was different for main app, widget, watch extension. When I save deviceIdentifier in appgroup userDefaults instead of sharedUserDefaults and use it from appgroup userDefaults - i get solution - no dublicate actions. What are you think about it?
Can you add capability to use appgroup userDefaults? Or is it wrong?

**Comment by mentrena** [comment]
Hi @shenmueXVI , thanks for the report. I don't have much experience with WatchKit, could you help clarify the following:

So, by using an app group, your WatchKit extension and main app basically use the same Core Data stack at the same time?
If you are using an Apple Watch with iPhone A, is it possible at some point to switch the same Apple Watch to be used with a different iPhone B? Where would the Core Data stack reside in that case?

Sorry if the question doesn't make sense, but it might change the approach to follow for the solution :)

**Comment by shenmueXVI** [comment]
OK. I have iPhone A. iPhone A has my app. My app has main app, watchKit Extension, Today Extension. They use common CoreData container. CoreData container located in common app group path, for example:

`NSString* storeName = @"nameBD.sqlite";
NSURL* groupURL = [[NSFileManager defaultManager] containerURLForSecurityApplicationGroupIdentifier:@"nameAppGroup"];
NSURL* storeURL = [groupURL URLByAppendingPathComponent:storeName];`

So, main app, watch extension, today extension use common CoreData container in iPhone A. Also main app, watch extension, today extension use common cloudKit container in iPhone A. And main app, watch extension, today extension use method synchronizeWithCompletion: from your library.

I think there is probem, because iPhone A has one application (main app, watch extension, today extension) with common CoreData container, but deviceIdentifier (from library) for main app, watch extension, today extension will be different. And after iCloud sync - I have duplicate data in coreData if will be changes.

When I use common deviceIdentifier in iPhone A for my application (main app, watch extension, today extension) - I get solution without dublicate data

In general, if I change CoreData in my main App, watchExtension or Today extension will be use same CoreData stack and reverse

Also I use QSPrimaryKey protocol in my model objects

**Comment by mentrena** [comment]
Ah right, the extension runs on the phone, not the watch. Thanks

**Comment by mentrena** [comment]
@shenmueXVI The solution seems easy but I would like to get to the root of the problem first. The use of `deviceIdentifier` is intended as an optimisation in the library; even if the different parts of your app use different `deviceIdentifier` that shouldn't lead to duplicates if you're providing primary keys for your objects 🤔

Even if you use a common CoreData container, are changes visible by the other device right away? For example, does your code need to do something so that changes you make on the Watch become visible in the main app, or does it happen immediately?

Would you have some sample code by chance?

**Comment by shenmueXVI** [comment]
Hm, I'm agree with you. I will check in more detail the logic of my code and report later about it

**Comment by mentrena** [comment]
@Kautenja, as I was saying here I'm not too familiar with how a WatchKit extension is architected.
When you use app groups, do you keep your store file (sqlite, I guess) in the app group, then create one managed object context in your main app and another context in your extension?

**Comment by Kautenja** [comment]
Yes the sqllite file moves from the default location to a "shared" location that the main application as well as any extensions have access to.

This subclass of `NSPersistentContainer` overrides the default directory to point to the shared location wherever that is within the OS.

```swift
/// A custom NSPersistentContainer for accessing the shared group. This enables
/// external processes to access the same core data stack objects.
/// This is utilized by today extensions, notification extensions, and the watch
/// app if any
class GroupPersistentContainer: NSPersistentContainer {
    
    /// the id of the shared group for the applications
    static let groupID = "group.appIdentifier"
    
    /// Return the default directory for the group.
    /// - returns: the URL to the app group for the apps/extensions
    override class func defaultDirectoryURL() -> URL{
        return FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: groupID)!
    }
    
}
```

I then use a singleton pattern to create  a class for housing the new persistent container and sync kit machinery. The relevant code for that looks like this.

```swift
/// The central core data stack for the application, extensions, and apple watch
/// application
class CoreDataStack: NSObject {
    
    /// The private constructor for the singleton instance
    private override init() {
        super.init()
        // load the view context to avoid some weird startup problems
        _ = persistentContainer.viewContext
    }
    
    /// the private shared instance of the stack
    private static var singleton: CoreDataStack?
    
    /// the public accessor for the shared instance
    public static var shared: CoreDataStack {
        guard let _shared = singleton else {
            singleton = CoreDataStack()
            return singleton!
        }
        return _shared
    }
    
    /// The persistent container for the application. This implementation
    /// creates and returns a container, having loaded the store for the
    /// application to it. This property is optional since there are legitimate
    /// error conditions that could cause the creation of the store to fail.
    lazy var persistentContainer: NSPersistentContainer = {
        let container = GroupPersistentContainer(name: "_Ring")
        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
             // error handling
        })
        return container
    }()

    // ... 

}
```

I can then access the shared instance of the stack from any target to perform insertions, updates, deletions, etc.

### TL;DR

* one sqllite file in a shared folder in the OS
* each target has it's own copy of the core data stack pointing to the same sqllite file

**Comment by spotlessicode** [comment]
That would be a great solution, @shenmueXVI and @Kautenja this app group solution works with CoreData sharing at Watch OS 3 too? As I read, since Watch OS 2 IOS and Watch Kit has separate CoreData, you can't share frameworks and you can't use the same MOC on Watch and IOS, that means can't send or reach the default objects from IOS, because they exist just on that MOC. Therefore I wanted to start to use CloudKit to download CKRecords from the same container and set the separate CoreData stack (WatchKit own CoreData with own Object Classes) on the WatchKit side. I would appreciate it if you would confirm that it really works. 

> You can share code, but not frameworks, between your iOS app and Watch app. Because the apps run on separate platforms with different architectures, source files must be compiled separately for each platform. If you still want to use a framework to manage any shared source files, you must create separate framework targets for each platform and add your shared source files to each framework.
> 
> If you already have an iOS framework, you can duplicate the framework and modify it to support watchOS.
> 
> After you create a separate framework target for your WatchKit extension, you must manage the contents of that framework separately for both iOS and watchOS. If you add files to the iOS version of the framework, remember to add them to the watchOS version.

[from the documentation](https://developer.apple.com/library/content/documentation/General/Conceptual/WatchKitProgrammingGuide/iOSSupport.html#//apple_ref/doc/uid/TP40014969-CH21-SW1)

But Apple [Suggest WatchConencttivity, and background tasks](https://developer.apple.com/library/content/documentation/General/Conceptual/WatchKitProgrammingGuide/SharingData.html#//apple_ref/doc/uid/TP40014969-CH29-SW1), I have that solution now, but it has limitations, so you can't send the whole database with [updateApplicationContext:error:](https://developer.apple.com/documentation/watchconnectivity/wcsession/1615621-updateapplicationcontext). the other option for the rest of the data is [transferUserInfo:](https://developer.apple.com/documentation/watchconnectivity/wcsession/1615671-transferuserinfo), but this deliver data "when the time is good for that.. battery charge and so on"... 

**My approach:**
But since Watch OS 3 direct CloudKit support is available, which would be the best for long-term. That means too, that you don't need to use app group anymore, just set the same CoreData model at the WatchKit Extension and download data from CloudKit as on IOS side. For that, you should put SyncKit libraries into WathKit Extension, and add SyncKit framework to WathcKit too.  - as @manuelentrena has already written, for that it should be updated for the latest CloudKit API... but I would like to make a test, put the library manually into it, and overwrite codes what necessary I'll notify you about the result. 

> iCloud interactions. Starting with watchOS 3, the WatchKit extension can communicate directly with CloudKit and other iCloud technologies.

**Comment by mentrena** [comment]
I'm concerned that the solution might not be as simple.
When you initialise the synchronizer you pass it a NSManagedObjectContext, and it will track changes in that context, and it will do so by using its own internal CoreData stack --see `QSCloudKitSynchronizer+CoreData.m`

In your case, there will be only one copy of your database, but there's two contexts, two synchronizers, and two internal databases with tracking information. So when your extension downloads a record from iCloud and inserts a new object in CoreData, the synchroniser in the main app detects the insertion and creates new tracking data for it, then tries to upload the object to iCloud --therefore creating a duplicate object.
At this point, if the two synchronizers were using the same `deviceIdentifier`, when the extension downloads the duplicate record from iCloud it won't save it, as it thinks it was created by itself so there's no point in saving the changes. However there might still be a duplicate record on iCloud. I don't get this though, since the two records would have tried to used the same CKRecordID (as it's derived from the object's entityType and primaryKey value), but one synchroniser didn't have the up-to-date  CKRecord metadata so it shouldn't have been able to upload anything and should have failed with an error. It's hard to fully know what's going on without reproducing the issue, don't you have a sample project with a watch kit extension and shared database in app group? 🤔

I think the right solution for this set up would require:
- Initialising the synchronizer with a persistenceStack that resides in the app group shared space.
- Updating `QSCoreDataChangeManager` to check if there's tracking data for a newly inserted object (as it might have been inserted by the extension)
- Using the same `deviceIdentifier` in both synchronizers.

**TL;DR**
I think it might not be as simple as using the same deviceIdentifier.

Alternatively, wouldn't it be possible to only synchronise on the main iOS app? The extension could request the main app to synchronise by using WatchConnectivity, while the database would still be in the app group so you get the changes in the extension.

Or, if the database is not too big, wouldn't it be possible to have separate databases in iOS app and extension and do all synchronisation through iCloud?

What are the limitations in these approaches?

**Comment by spotlessicode** [comment]
I meant - download data from the same CloudKit container and update Core Data stack on WatchKit. But not update CloudKit from Watch, instead of that, as Apple suggests, send changes with WatchConnectivity to IOS, and when IOS App launch saves changes there + sync to Cloud. Watch OS just should download that database, but not change it.

**Comment by Kautenja** [comment]
hmm no i haven't tried my solution on a watch app yet, but the code should behave exactly the same. It seems as if simple extensions like today and notification extensions behave differently than full fledged watch apps (which i haven't but at some point intend to implement in a project using SyncKit). I'll outline my use case to define some solutions and there respective problems:

### Use Case

Users will have access to Today extensions that can reduce the amount of time they spend in the application. These extensions will display time sensitive data that can potentially change on other devices while the main instance (the app) is not running. As such, the extension itself should connect to iCloud.

#### App Groups (no iCloud)

Using just app groups allows the extensions so access the same database as the app. If the data changes on another device while the app isn't running, there will be no way to sync the data without actually opening the app, making the today extension data incorrect until the next open.

#### iCloud (no AppGroups)

Using just iCloud solves issues with duplication across databases as each application and extension will have its own local database. The problem here is that another important use case for the app is that iCloud be functionality be _entirely_ optional. This also adds external network communication between processes on the same device which places a requirement on internet for the extensions to work. Subscriptions aren't allowed in extensions so the extension will have to do a greedy refresh everytime it loads much like how a sync might occur on application did launch with options. Ultimately this solution solves the issue and might work for someone, but sadly doens't fit my use case.

#### Hybrid Solution

Combining app groups and iCloud, extensions will be able to not only access the database quickly, but perform incremental updates to keep their own as well as the apps data up to date. This means the app doesnt need to be running for data updates to occur. The current problem here is this issue.

#### Notification Center

One last idea is to use notification center. The main app could post notifications when the data relevant to the extension changes. This doesnt address the sync problem though, i.e. if data changed on another device while the app wasnt running on this device, the app on this device would need to open before this extension could get notified of changes.

*****

Those are the solutions I've considered with some analysis. I'm privy to the 3rd (Hybrid Solution) for a use case involving today/notification extensions. For a watch app I'd lean towards the 2nd (iCloud only) as it seems like Apple wants the watch to behave almost like a standalone device despite being tethered to a phone? Makes sense as the watch apps are "native" now, but idk that's my initial impression.

**Comment by Kautenja** [comment]
I don't have a sample project with this functionality as I just integrated it into a beta production app. I can throw one together though if that would be helpful in diagnosing the problem. Most of the code already exists, the project would just need setup. @mentrena would that be helpful?

**Comment by mentrena** [comment]
@Kautenja I think that would be helpful.
In the meantime I've created a branch here: https://github.com/mentrena/SyncKit/tree/appGroupSupport
with a basic, _untested_ implementation :)

For a fresh install, try using
```objc
+ (QSCloudKitSynchronizer *)cloudKitSynchronizerWithContainerName:(NSString *)containerName managedObjectContext:(NSManagedObjectContext *)context changeManagerDelegate:(id<QSCoreDataChangeManagerDelegate>)delegate suiteName:(NSString *)suiteName;
```

If you're adding the extension to an app that was using SyncKit before, then first try copying the sqlite file from `[QSCloudKitSynchronizer storePath]` to `[QSCloudKitSynchronizer  storePathWithAppGroup:(NSString *)suiteName]`

**Comment by spotlessicode** [comment]
I made my approach, and SynKit works with WatchKit, it downloads the whole database without AppGroup! ;) WC can send changes back to IOS which save and sync again. I tested with a real app on real devices. [Check it](https://github.com/spotlessicode/SyncKitToWatchKit/blob/master/README.md)

**Comment by Kautenja** [comment]
@mentrena I've put together a sample project here https://github.com/Kautenja/iOSTodayExtensionUbiquitySampleProject. A few caveats:

* it uses carthage so make sure carthage is installed or the run script that copies frameworks will fail when you try to build.
* there are some funky things with entitlements as they move between developer accounts. The README for the project outlines some of these issues and how to resolve them.

Definitely open an issue if you run into any problems with the sample. Hopefully it will be of use. Thanks for your continuous efforts on this project!

**Comment by Kautenja** [comment]
### Correction to earlier statement

I earlier stated that today extensions could not register for CloudKit update subscriptions. This is actually not true, they can. A typo i made in the sample project revealed this to me. Not really relevant to the issue, but I thought I'd correct it just cuz

**Comment by mentrena** [comment]
@Kautenja that was very helpful, thanks.
I made small changes to your project to import SyncKit as a pod and to make debugging easier, and I was able to reproduce the issue.
I've also done a quick test of my changes in the appSupportGroup branch and so far it seems to work.
I've pushed the changes to https://github.com/mentrena/todayextensiontest . Could you take a look and let me know your thoughts?

**Comment by Kautenja** [comment]
@mentrena I just cloned to code and set it up and everything is looking great so far!I get no duplication of records between the extensions and the main app and the extension alone can successfully sync with iCloud with no problem. I've confirmed on both iOS10 and iOS11 with physical devices (an iphone 7 and ipad air 2). 

I'll integrate the changes in the `appGroupSupport` branch  into my production SyncKit project later tonight / tomorrow and report back with any problems, but it seems as if this issue is resolved.

Just to confirm, passing the app group ID to the initializer for `QSCloudKitSynchronizer` using `suiteName` is the only change necessary to integrate App Groups with SyncKit? i.e.

```swift
let appGroupID = "group.com.company.app"
let sync = QSCloudKitSynchronizer(
    ...
    suiteName: appGroupID)
```

**Comment by shenmueXVI** [comment]
I started to use QSCoreDataStack with appGroup container (instead local doucements container), and I got solution about duplicate records (my suggestion about deviceIdentifier sometimes worked, but appered new bugs) . I think to need to use QSCoreDataStack with appGroup container if core data has common app group container. I will to use local solution QSCoreDataStack with appGroup. @manuelentrena what are you think about to add this change to general version?

**Comment by mentrena** [comment]
@shenmueXVI glad to hear that. Yes, that is the idea, I will release a new version some time this week.
@Kautenja if you're using app groups for a new app, that would be all. If you're adding app groups to an existing app that was already using SyncKit then you would need to move the existing SyncKit database into the shared container -- `QSSynchronizer+CoreData.h` now exposes the relevant file paths.

**Comment by mentrena** [comment]
0.5.1 has been published with support for App Groups, it fixes a couple of extra issues about app groups. Thanks both.
